parsec/mca/sched/llp/sched_llp_module.c

```diff
@@ -0,0 +1,656 @@
+ *
+ * @details
+ *    Creates a LIFO per execution stream, store it into es->scheduling_object, and
+ *    synchronize with the other execution streams using the barrier
+ *
+ *  @param[INOUT] es      the calling execution stream
+ *  @param[INOUT] barrier the barrier used to synchronize all the es
+ *  @return PARSEC_SUCCESS in case of success, a negative number otherwise
+ */
+static int flow_llp_init(parsec_execution_stream_t* es, struct parsec_barrier_t* barrier)
+{
+    /* Every flow creates its own local object */
+    parsec_lifo_with_prio_t *lifo = PARSEC_OBJ_NEW(parsec_lifo_with_prio_t);
+    es->scheduler_object = lifo;
+
+    /* All local allocations are now completed. Synchronize with the other
+     threads before setting up the entire queues hierarchy. */
+    parsec_barrier_wait(barrier);
+
+#if defined(PARSEC_PAPI_SDE)
+    if( 0 == es->th_id ) {
+        char event_name[PARSEC_PAPI_SDE_MAX_COUNTER_NAME_LEN];
+        snprintf(event_name, PARSEC_PAPI_SDE_MAX_COUNTER_NAME_LEN, "SCHEDULER::PENDING_TASKS::QUEUE=%d::SCHED=LLP", es->virtual_process->vp_id);
+        parsec_papi_sde_register_fp_counter(event_name, PAPI_SDE_RO|PAPI_SDE_INSTANT,
+                                     PAPI_SDE_int, (papi_sde_fptr_t)parsec_lifo_with_local_counter_length, es->virtual_process);
+        parsec_papi_sde_add_counter_to_group(event_name, "SCHEDULER::PENDING_TASKS", PAPI_SDE_SUM);
+        parsec_papi_sde_add_counter_to_group(event_name, "SCHEDULER::PENDING_TASKS::SCHED=LLP", PAPI_SDE_SUM);
+    }
+#endif
+    return PARSEC_SUCCESS;
+}
+
+/**
+ * @brief
+ *   Selects a task to run
+ *
+ * @details
+ *   Take the head of the calling execution stream LIFO as the selected task;
+ *   if that LIFO is empty, iterate over all other execution streams LIFOs,
+ *   using the eu_id as an index (modulo the number of execution streams in this
+ *   virtual process).
+ *
+ *   @param[INOUT] es     the calling execution stream
+ *   @param[OUT] distance the distance of the selected task. We return here
+ *                        how many LIFOs that are empty were tried
+ *   @return the selected task
+ */
+static parsec_task_t* sched_llp_select(parsec_execution_stream_t *es,
+                                      int32_t* distance)
+{
+    parsec_task_t *task = NULL;
+    parsec_lifo_with_prio_t *es_sched_obj = (parsec_lifo_with_prio_t*)es->scheduler_object;
+    int i, d = 0;
+
+    task = (parsec_task_t*)parsec_lifo_pop(&es_sched_obj->lifo);
+
+    if (NULL == task) {
+        for(i = (es->th_id + 1) % es->virtual_process->nb_cores;
+            i != es->th_id;
+            i = (i+1) % es->virtual_process->nb_cores) {
+            d++;
+            parsec_lifo_with_prio_t *sched_obj;
+            sched_obj = (parsec_lifo_with_prio_t*)es->virtual_process->execution_streams[i]->scheduler_object;
+            task = (parsec_task_t*)parsec_lifo_pop(&sched_obj->lifo);
+            if( NULL != task ) {
+                *distance = d;
```

Comment:
you can also compute the distance without a temporary variable. Example:`*distance = (i - es->th_id + es->virtual_process->nb_cores) % es->virtual_process->nb_cores;`